lib/network_utils.ml

```diff
@@ -1163,3 +1200,196 @@ module Ethtool = struct
+				let driver_options = Re.split (Re_perl.compile_pat "[ \t]+") trimed_s in
+				List.map parse_driver_options driver_options
+				|> String.concat " "
+			else
+				trimed_s
+		in
+		let lines = try Unixext.read_lines file_path with _ -> [] in
+		let new_conf = List.map parse_single_line lines in
+		!has_probe_conf, !need_rebuild_initrd, new_conf
+
+	(*
+	returns ( a * b * c) where
+	a indicates the probe configuration already has the max_vfs options, meaning the device doesn't support sysfs and will be configed by modprobe
+	b indicates some changes shall be made on the coniguration to enable sriov to max_vfs, so we shall rebuild the initrd.
+	and c is the configurations after these changes
+	*)
+	let parse_modprobe_conf driver max_vfs =
+		let file_path = Printf.sprintf "/etc/modprobe.d/%s.conf" driver in
+		let option = gen_options_for_maxvfs driver max_vfs in
+		parse_modprobe_conf_internal file_path driver option
+
+	let get_sriov_info lspci_output =
+		if Re.execp (Re_perl.compile_pat "Capabilities:.+SR-IOV") lspci_output then
+		begin
+			let group = Re.exec_opt (Re_perl.compile_pat "Total VFs: (\d+), Number of VFs: (\d+)") lspci_output in
+			match group with
+			| None -> {max_vfs=0; num_vfs=0}
+			| Some x -> {max_vfs = (int_of_string (Re.Group.get x 1)) - 1;
+				num_vfs = (int_of_string (Re.Group.get x 2))}
+		end
+		else
+			{max_vfs=0; num_vfs=0}
+
+	let get_lspci_output dev = 
+		call_script lspci ["-vv"; "-s"; (Sysfs.get_pcibuspath dev)]
+
+	let get_capabilities dev = 
+		try
+			let info = get_lspci_output dev |> get_sriov_info in
+			if info.max_vfs = 0 then [] else ["sriov"]
+		with _ -> []
+
+	let set_sriov_via_sysfs dev num_vfs =
+		let interface = Printf.sprintf "/sys/class/net/%s/device/sriov_numvfs" dev in
+		let oc = open_out interface in
+		let check_vf_numbers dev num_vfs = 
+			try
+				let info = get_lspci_output dev |> get_sriov_info in
+				info.num_vfs = num_vfs 
+			with _ -> false
+		in
+		try
+			Pervasiveext.finally (fun () ->
+				output_string oc (string_of_int num_vfs))
+				(fun () -> close_out oc);
+			if check_vf_numbers dev num_vfs then Ok Sysfs_successful
+			else begin
+					debug "%s: no exceptions but fail to set SRIOV" dev;
+					Error (Unknown "unable to enable via sysfs")
+			end
+		with
+		| Sys_error s when Xstringext.String.has_substr s "out of range of" -> Error Bus_out_of_range
+		| Sys_error s when Xstringext.String.has_substr s "not enough MMIO resources" -> Error Not_enough_mmio_resources
+		| e ->
+			debug "%s: set sriov error with exception %s" dev (Printexc.to_string e);
+			Error (Unknown "unable to enable via sysfs")
+
+	let get_driver dev =
+		match (Sysfs.get_driver_name dev) with
+		| Some driver -> driver
+		| None -> raise (Sys_error ("cannot get driver name for " ^ dev))
+
+	let write_conf_file driver content=
+		Unixext.write_string_to_file (Printf.sprintf "/etc/modprobe.d/%s.conf" driver) (String.concat "\n" content)
+
+	let enable_internal dev =
```

Comment:
See the comment on `make_vf_conf_internal` below. Most of this function should be moved into the `enable` function in network_server.ml.